core/arch/riscv/kernel/sbi.c

```diff
@@ -42,3 +48,16 @@ int sbi_boot_hart(uint32_t hart_id, paddr_t start_addr, unsigned long arg)
 
 	return ret.error;
 }
+
+void __noreturn __naked teeabi_return_to_ree(unsigned long arg0,
```

Comment:
LGTM but I have a question: why `naked`?

Reply:
This function is unnecessary to operate stack (function prologue/epilogue) because we are yielding control to REE.
Therefore `naked` attribute is added.

Reply:
Seems like a strange optimization. Given that you already have the `__noreturn` attribute, does it make much difference?

Reply:
> Given that you already have the __noreturn attribute, does it make much difference?

When I built the function with `__noreturn`, it only removes the function epilogue.

If only `__noreturn` attribute is applied, the disassembler shows:
```
000000008e000906 <teeabi_return_to_ree>:
teeabi_return_to_ree():
    8e000906:	715d                	add	sp,sp,-80
    8e000908:	e486                	sd	ra,72(sp)
    8e00090a:	e0a2                	sd	s0,64(sp)
    8e00090c:	0880                	add	s0,sp,80
    8e00090e:	fca43c23          	sd	a0,-40(s0)
    8e000912:	fcb43823          	sd	a1,-48(s0)
    8e000916:	fcc43423          	sd	a2,-56(s0)
    8e00091a:	fcd43023          	sd	a3,-64(s0)
    8e00091e:	fae43c23          	sd	a4,-72(s0)
    8e000922:	faf43823          	sd	a5,-80(s0)
    8e000926:	fd843503          	ld	a0,-40(s0)
    8e00092a:	fd043583          	ld	a1,-48(s0)
    8e00092e:	fc843603          	ld	a2,-56(s0)
    8e000932:	fc043683          	ld	a3,-64(s0)
    8e000936:	fb843703          	ld	a4,-72(s0)
    8e00093a:	fb043783          	ld	a5,-80(s0)
    8e00093e:	4801                	li	a6,0
    8e000940:	005448b7          	lui	a7,0x544
    8e000944:	54588893          	add	a7,a7,1349 # 544545 <__flatmap_rw_size+0x3d1545>
    8e000948:	00000073          	ecall
    8e00094c:	87aa                	mv	a5,a0
    8e00094e:	873e                	mv	a4,a5
    8e000950:	87ae                	mv	a5,a1
    8e000952:	fee43023          	sd	a4,-32(s0)
    8e000956:	fef43423          	sd	a5,-24(s0)
    8e00095a:	4681                	li	a3,0
    8e00095c:	4601                	li	a2,0
    8e00095e:	4581                	li	a1,0
    8e000960:	4501                	li	a0,0
    8e000962:	4aa150ef          	jal	8e015e0c <__do_panic>
```

When `__noreturn` and `__naked` attributes are applied, the disassembler shows:
```
000000008e000906 <teeabi_return_to_ree>:
teeabi_return_to_ree():
    8e000906:	4801                	li	a6,0
    8e000908:	005448b7          	lui	a7,0x544
    8e00090c:	54588893          	add	a7,a7,1349 # 544545 <__flatmap_rw_size+0x3d1545>
    8e000910:	00000073          	ecall
    8e000914:	87aa                	mv	a5,a0
    8e000916:	873e                	mv	a4,a5
    8e000918:	87ae                	mv	a5,a1
    8e00091a:	fee43823          	sd	a4,-16(s0)
    8e00091e:	fef43c23          	sd	a5,-8(s0)
    8e000922:	4681                	li	a3,0
    8e000924:	4601                	li	a2,0
    8e000926:	4581                	li	a1,0
    8e000928:	4501                	li	a0,0
    8e00092a:	488150ef          	jal	8e015db2 <__do_panic>
```
Where function prologue and epilogue are both removed.

Reply:
If you care about such things, why not implement it in assembly?
However, if this is how you prefer to do it in riscv I will not interfere.

Reply:
Thanks, I decided to implement it in assembly